### PR TITLE
365 make removevalidator world callable

### PR DESCRIPTION
**Pull request by glottologist**
I ran out of time to finish #365 off entirely but this PR is for the small headway I made into the ticket.

### CODE REVIEW

**Comment by ruuda** [comment]
I changed the base branch to `unstake` so the changes from #373 don’t show up in the diff here. GitHub will change it to `next` automatically when #373 is merged into `next`.

File: program/src/process_management.rs [hunk 1]
```diff
@@ -84,9 +84,11 @@ pub fn process_remove_validator(
...
-    let removed_validator = lido
-        .validators
-        .remove(accounts.validator_vote_account_to_remove.key)?;
+    let removed_validator = lido.validators.remove_on(
+        accounts.validator_vote_account_to_remove.key,
+        |x| x.can_be_removed(),
+        LidoError::ValidatorIsStillActive,
+    )?;
 
     if removed_validator.fee_credit != StLamports(0) {
```
Comment (ruuda) [inline program/src/process_management.rs:93]: Maybe for defensive programming, we can assert that `validator.stake_accounts_balance == Lamports(0)` before we remove the validator. If it has no stake accounts, then its stake accounts balance *should* be 0, but if for whatever reason that is not the case, it is probably better to fail the transaction and keep the validator so we can investigate, than to allow it to succeed and possibly lose track of some SOL.

File: program/tests/tests/add_remove_validator.rs [hunk 2]
```diff
@@ -43,11 +48,57 @@ async fn test_remove_validator_without_unclaimed_credits() {
...
 
-    context
-        .try_remove_validator(vote_account)
-        .await
-        .expect("Failed to remove validator.");
+    let result = context.try_remove_validator(vote_account).await;
 
-    let solido = context.get_solido().await;
-    assert_eq!(solido.validators.len(), 0);
+    assert_solido_error!(result, LidoError::ValidatorIsStillActive);
```
Comment (ruuda) [inline program/tests/tests/add_remove_validator.rs:53]: The test name is misleading now, removing a an _inactive_ validator without unclaimed credits should succeed ... if it fails on being active, then the fact that it has no unclaimed credits isn’t really relevant any more. Also, there is no test now that tests successful removal of the validator. I think we can extend this test to

* After asserting that it has no unclaimed fees, deactivate the validator.
* Then remove it, and assert that that succeeds.

And then we can add a new test to confirm that trying to remove an active validator fails.

File: program/tests/tests/add_remove_validator.rs [hunk 3]
```diff
@@ -43,11 +48,57 @@ async fn test_remove_validator_without_unclaimed_credits() {
...
+    );
+
+    // This was also the first deposit, so that should have created one stake account.
+    assert_eq!(validator_after.stake_seeds.stake_accounts_seed_begin, 0);
+    assert_eq!(validator_after.stake_seeds.stake_accounts_seed_end, 1);
+
+    let result = context.try_remove_validator(validator.vote_account).await;
+
+    // The validator should not be able to be removed if it is still active
+    //  (i.e. the active flag is set toe true OR it has stake accounts)
```
Comment (ruuda) [inline program/tests/tests/add_remove_validator.rs:102]: s/toe/to/

File: program/tests/tests/add_remove_validator.rs [hunk 4]
```diff
@@ -43,11 +48,57 @@ async fn test_remove_validator_without_unclaimed_credits() {
...
+    assert_solido_error!(result, LidoError::ValidatorIsStillActive);
```
Comment (ruuda) [inline program/tests/tests/add_remove_validator.rs:103]: If removal fails on the fact that the validator is still active, then all of this business above of giving it stake was not relevant. We should also deactivate the validator (which does not deactivate the stake), and then we can test that it’s really the stake that prevents removing. The function to deactivate is here: https://github.com/chorusone/solido/blob/295a263525fb20c0e17c01bae994c76dd74c4d12/program/tests/context.rs#L707

**Review by ruuda — COMMENTED** [review]
I have a few comments about how to finish this, but I suppose you will not get to it @glottologist?

**Comment by enriquefynn** [comment]
I just rebased on top of `unstake`, @glottologist can work on the issues @ruuda raised.

File: program/src/state.rs [hunk 5]
```diff
@@ -673,6 +673,32 @@ impl Default for Validator {
...
+    pub fn has_no_stake_accounts(&self) -> bool {
+        self.stake_seeds.begin == self.stake_seeds.end
+    }
+
+    pub fn check_can_be_removed(&self) -> Result<(), LidoError> {
+        if self.active {
+            msg!("Refusing to remove validator because it is still active, deactivate it first.");
+            return Err(LidoError::ValidatorIsStillActive);
+        }
+        if !self.has_no_stake_accounts() {
```
Comment (ruuda) [inline program/src/state.rs:686]: We can simplify the double negation by having a method `has_stake_accounts` that checks `begin != end`.

File: program/tests/tests/add_remove_validator.rs [hunk 6]
```diff
@@ -34,24 +38,30 @@ async fn test_successful_add_validator() {
...
+async fn test_removing_validator_with_stake_accounts_should_fail() {
+    let (mut context, _) = Context::new_with_two_stake_accounts().await;
+    let validator = &context.get_solido().await.validators.entries[0];
+    let result = context.try_remove_validator(validator.pubkey).await;
+
+    // The validator should not be able to be removed if it is still active
+    //  (i.e. the active flag is set toe true OR it has stake accounts)
+    assert_solido_error!(result, LidoError::ValidatorIsStillActive);
+}
+
```
Comment (ruuda) [inline program/tests/tests/add_remove_validator.rs:64]: We should also add a third test where the validator has no stake accounts, but it does have unclaimed credits, and test that we refuse to remove it in that case too.

**Review by ruuda — COMMENTED** [review]
One test is still missing, and I have one optional suggestion. LGTM aside from those!

File: program/tests/tests/add_remove_validator.rs [hunk 7]
```diff
@@ -51,6 +51,51 @@ async fn test_successful_remove_validator() {
...
+    let solido = context.get_solido().await;
+    let vote_account = solido.validators.entries[0].pubkey;
+    assert_eq!(
+        solido.validators.entries[0].entry.fee_credit,
+        StLamports(62_301_530_769)
+    );
+
+    context.deactivate_validator(vote_account).await;
+    // let solido = context.get_solido().await;
+    // let vote_account = solido.validators.entries[0].pubkey;
```
Comment (ruuda) [inline program/tests/tests/add_remove_validator.rs:94]: Can this be removed?

File: program/src/process_management.rs [hunk 8]
```diff
@@ -84,9 +84,11 @@ pub fn process_remove_validator(
     let mut lido = deserialize_lido(program_id, accounts.lido)?;
     lido.check_manager(accounts.manager)?;
```
[the remark this answers is not in the document]
Comment (ruuda) [reply program/src/process_management.rs]: > This needs a change in instruction.rs to remove the manager from the input accounts as well.

We forgot this part ... Discovered by Neodyme, will fix.

**Pull request merged**